pallets/token-mux/src/lib.rs

```diff
@@ -0,0 +1,378 @@
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+
+pub use cfg_traits::TokenSwaps;
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use cfg_traits::{OrderRatio, TokenSwaps};
+	use cfg_types::{orders::OrderInfo, tokens::CustomMetadata};
+	use frame_support::{
+		pallet_prelude::{DispatchResult, *},
+		traits::{
+			fungibles,
+			fungibles::Mutate,
+			tokens::{Fortitude, Precision, Preservation},
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::{OriginFor, *};
+	use orml_traits::asset_registry::{self, Inspect as _};
+	use sp_arithmetic::{
+		traits::{AtLeast32BitUnsigned, EnsureDiv, EnsureMul},
+		FixedPointOperand,
+	};
+	use sp_runtime::traits::{AccountIdConversion, EnsureFixedPointNumber, One};
+
+	use super::*;
+
+	pub type BalanceOf<T> = <<T as Config>::Tokens as fungibles::Inspect<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// The source of truth for the existence and potential local
+		/// representation of assets.
+		type AssetRegistry: asset_registry::Inspect<
+			CustomMetadata = CustomMetadata,
+			AssetId = Self::CurrencyId,
+		>;
+
+		/// Balance type for incoming values
+		type BalanceIn: Member
+			+ Parameter
+			+ FixedPointOperand
+			+ AtLeast32BitUnsigned
+			+ EnsureMul
+			+ EnsureDiv
+			+ MaxEncodedLen
+			+ Into<BalanceOf<Self>>
+			+ From<BalanceOf<Self>>;
+
+		/// Balance type for outgoing values
+		type BalanceOut: Member
+			+ Parameter
+			+ FixedPointOperand
+			+ AtLeast32BitUnsigned
+			+ EnsureMul
+			+ EnsureDiv
+			+ MaxEncodedLen
+			+ Into<BalanceOf<Self>>
+			+ From<BalanceOf<Self>>;
+
+		/// Type for price ratio for cost of incoming currency relative to
+		/// outgoing
+		type BalanceRatio: Parameter
+			+ Member
+			+ sp_runtime::FixedPointNumber
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		/// The token swap order identifying type
+		type OrderId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord + MaxEncodedLen;
+
+		/// The general asset type
+		type CurrencyId: Parameter
+			+ Member
+			+ Copy
+			+ MaxEncodedLen
+			+ From<Self::LocalAssetId>
+			+ TryInto<Self::LocalAssetId>;
+
+		/// The local asset type
+		type LocalAssetId: From<cfg_types::tokens::LocalAssetId>;
+
+		/// The type for handling transfers, burning and minting of
+		/// multi-assets.
+		type Tokens: fungibles::Inspect<Self::AccountId, AssetId = Self::CurrencyId>
+			+ Mutate<Self::AccountId, AssetId = Self::CurrencyId>;
+
+		/// The type for retrieving and fulfilling swap orders.
+		type Swaps: TokenSwaps<
```

Comment:
Super NIT, I would call this `OrderBook` (actually, I would call `TokenSwaps` as `OrderBook`, but that for another PR). This does not perform "swaps", it performs actions over an order book. The difference is now more clear with `pallet-swaps`, which does use swaps.

Reply:
@wischli not sure if you read this or it has become too many collapsed 😆

Reply:
Thanks for the reminder. I agree and will change it!